src/Tests/RedirectUITest.php

```diff
@@ -386,4 +391,27 @@ public function assertRedirect($path, $expected_ending_url, $expected_ending_sta
     $this->assertEqual($headers[0][':status'], $expected_ending_status);
   }
 
+  /**
+   * Test cache tags.
+   *
+   * @todo Not sure this belongs in a UI test, but a full web test is needed.
+   */
+  public function testCacheTags() {
+    // Enable internal page cache.
+    $this->container->get('config.factory')->getEditable('system.performance')
+      ->set('cache.page.use_internal', TRUE)
+      ->set('cache.page.max_age', 300)
+      ->save();
+
+    /** @var \Drupal\redirect\Entity\Redirect $redirect1 */
+    $redirect1 = $this->storage->create();
+    $redirect1->setSource('node');
+    $redirect1->setRedirect('admin');
+    $redirect1->setStatusCode(301);
+    $redirect1->save();
+
+    $this->drupalHead('node');
+    $this->assertCacheTag($redirect1->getCacheTags());
```

Comment:
This however is the part you have to do yourself, because this happens in the HtmlRenderer by default, and we're not going through that. 

So in onKernelRequestCheckRedirect(), where the response object is created, you need to add the cache tag header like that.